pkg/controller/restore-operator/backup_handler.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2017 The etcd-operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"k8s.io/client-go/kubernetes"
+
+	api "github.com/coreos/etcd-operator/pkg/apis/etcd/v1beta2"
+	"github.com/coreos/etcd-operator/pkg/backup"
+	"github.com/coreos/etcd-operator/pkg/backup/backend"
+	"github.com/coreos/etcd-operator/pkg/backup/backupapi"
+	backupS3 "github.com/coreos/etcd-operator/pkg/backup/s3"
+	"github.com/coreos/etcd-operator/pkg/cluster/backupstorage"
+	"github.com/sirupsen/logrus"
+)
+
+const tmpdir = "/tmp"
+
+// backupServerCloser wraps around backup.BackupServer with a close method
+// that cleans up resources associated with backup.BackupServer.
+type backupServerCloser struct {
+	*backup.BackupServer
+	awsDir    string
+	backupDir string
+}
+
+// close removes aws credential/config dir and backupDir dir which were
+// created for BackupServer.
+func (bsc *backupServerCloser) close() {
+	os.RemoveAll(bsc.awsDir)
+	os.RemoveAll(bsc.backupDir)
+}
+
+func (r *Restore) handleBackup(spec *api.BackupSpec) error {
+	var (
+		bsc *backupServerCloser
+		err error
+	)
+	switch spec.StorageType {
+	case api.BackupStorageTypeS3:
+		bsc, err = createS3BackupServer(r.kubecli, spec.S3, r.namespace, spec.ClusterName)
+		if err != nil {
+			return err
+		}
+	default:
+		logrus.Fatalf("unknown StorageType: %v", spec.StorageType)
+	}
+	r.mu.Lock()
+	r.backupServers[spec.ClusterName] = bsc
+	r.mu.Unlock()
+	return nil
+}
+
+func createS3BackupServer(kubecli kubernetes.Interface, s3 *api.S3Source, namespace, clusterName string) (*backupServerCloser, error) {
+	awsDir, err := ioutil.TempDir(tmpdir, "")
```

Comment:
Duplicate to backup-operator/s3_backup.go ?
WIP Anyway. Will look again at final.